### PR TITLE
advertise transport encoding in http transfers to

### PR DESCRIPTION
**Pull request by Mic92**
tl;dr: With this 1 line change I was able to get a speedup of 1.5x on 1Gbit/s wan connections by enabling zstd compression in nginx.

Also nix already supported all common compression format for http transfer, webservers usually only enable them if they are advertised through the Accept-Encoding header.

This pull requests makes nix advertises content compression support for zstd, br, gzip and deflate.

It's particular useful to add transparent compression for binary caches that serve packages from the host nix store in particular nix-serve, nix-serve-ng and harmonia.

I tried so far gzip, brotli and zstd, whereas only zstd was able to bring me performance improvements for 1Gbit/s WAN connections.

The following nginx configuration was used in combination with the [zstd module](https://github.com/tokers/zstd-nginx-module) and [harmonia](https://github.com/nix-community/harmonia/)

```nix
{
  services.nginx.virtualHosts."cache.yourhost.com" = {
    locations."/".extraConfig = ''
      proxy_pass http://127.0.0.1:5000;
      proxy_set_header Host $host;
      proxy_redirect http:// https://;
      proxy_http_version 1.1;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;

      zstd on;
      zstd_types application/x-nix-archive;
    '';
  };
}
```

For testing I unpacked a linux kernel tarball to the nix store using this command `nix-prefetch-url --unpack https://cdn.kernel.org/pub/linux/kernel/v6.x/linux-6.1.8.tar.gz`.

Before:

```console
$ nix build && rm -rf /tmp/hello  && time ./result/bin/nix copy --no-check-sigs --from https://cache.thalheim.io --to 'file:///tmp/hello?compression=none' '/nix/store/j42mahch5f0jvfmayhzwbb88sw36fvah-linux-6.1.8.tar.gz'
warning: Git tree '/scratch/joerg/nix' is dirty

real    0m18,375s
user    0m2,889s
sys     0m1,558s
```

After:

```console
$ nix build && rm -rf /tmp/hello  && time ./result/bin/nix copy --no-check-sigs --from https://cache.thalheim.io --to 'file:///tmp/hello?compression=none' '/nix/store/j42mahch5f0jvfmayhzwb
b88sw36fvah-linux-6.1.8.tar.gz'

real    0m11,884s
user    0m4,130s
sys     0m1,439s
```

Signed-off-by: Jörg Thalheim <joerg@thalheim.io>

# Motivation

# Context

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

### CODE REVIEW

File: src/libstore/filetransfer.cc [hunk 1]
```diff
@@ -101,6 +101,7 @@ struct curlFileTransfer : public FileTransfer
                     this->result.data.append(data);
               })
         {
+            requestHeaders = curl_slist_append(requestHeaders, "Accept-Encoding: zstd, br, gzip, deflate");
```
Comment (SuperSandro2000) [inline src/libstore/filetransfer.cc:104]: We can't easily get this list from curl or at compile time?

Comment (Mic92) [reply src/libstore/filetransfer.cc:104]: @SuperSandro2000 this is actually not curl decompressing but nix has its own decompression sinks below. In fact my first version was just doing `curl_easy_setopt(curl, CURLOPT_ACCEPT_ENCODING, "");` and disable having nix doing the decompression afterwards (or else it would break).

Comment (thufschmitt) [reply src/libstore/filetransfer.cc:104]: Except from `brotli`, the decompression is transparently handled by `libarchive` IIRC. So ideally we could just query `libarchive` for that, but after a quick search I couldn't find a way to do that.

**Comment by SuperSandro2000** [comment]
> webservers usually only enable them if they are advertised through the Accept-Encoding header.

They must only send what is written in Accept-Encoding. Thats why in addition to zstd supporting at least brotli and usually gzip, too is necessary for a normal webserver. Most browsers don't have adopted zstd yet or only behind a flag. Not that relevant for us here since nix support brotli for a while.

File: src/libstore/filetransfer.cc (hunk 1 above)
Comment (Mic92) [reply src/libstore/filetransfer.cc:104]: @thufschmitt the alternative would be to let curl handle all of the decompression. Just make a call which way do you prefer.

Comment (thufschmitt) [reply src/libstore/filetransfer.cc:104]: Dunno, I have hardly ever touched these parts of the codebase, and I've no idea why things are this way :man_shrugging: I guess @edolstra would have a more informed opinion here.

What would be the cost of letting `curl` handle it? Should it just be a matter of setting `CURLOPT_ACCEPT_ENCODING` or is there something more involved?

Comment (edolstra) [reply src/libstore/filetransfer.cc:104]: From what I remember, curl doesn't support all the formats that Nix supports (e.g. bzip2, and I think the curl in nixpkgs didn't have brotli enabled at the time). And Nix has to have its own support for these formats anyway. So there isn't much advantage to letting curl handle the decompression.

Comment (SuperSandro2000) [reply src/libstore/filetransfer.cc:104]: > I think the curl in nixpkgs didn't have brotli enabled at the time

I changed that a good while ago and moved brotili support from curlFull to curl.

So we can just keep this as is and maybe need to update the values manual in the future due to the lack of better options?

**Comment by NickCao** [comment]
My earlier attempt at this: https://github.com/NixOS/nix/pull/6694

**Comment by Mic92** [comment]
Great @NickCao has implemented the other approach in https://github.com/NixOS/nix/pull/6694, where nix offloads decompressing to libcurl. This means both variants are implemented and you just need to merge the one that you prefer.

File: src/libstore/filetransfer.cc (hunk 1 above)
Comment (thufschmitt) [inline src/libstore/filetransfer.cc:104]:
```suggestion
            requestHeaders = curl_slist_append(requestHeaders, "Accept-Encoding: zstd, br, gzip, deflate, bzip2, xz");
```

I think we support at least these two (though I've no idea whether adding them is useful :man_shrugging: )

**Review by thufschmitt — APPROVED** [review]
> both variants are implemented and you just need to merge the one that you prefer. I think both are fine.

I actually think we can just merge that one since it's just a one-liner change and then think whether we want the full-curl approach.

So modulo a small suggestion, I think this is good to go

File: src/libstore/filetransfer.cc (hunk 1 above)
Comment (Mic92) [reply src/libstore/filetransfer.cc:104]: `xz` might be, but I am not aware of webserver implementations for that.

**Comment by Mic92** [comment]
Squashed.

**Review by tomberek — APPROVED** [review]
